<!DOCTYPE html>
<html lang='en'>
<head>
<title>bpo-30353.patch\python3\main - aports - Main aports tree
</title>
<meta name='generator' content='cgit v1.1'/>
<meta name='robots' content='index, nofollow'/>
<link rel='stylesheet' type='text/css' href='/cgit.css'/>
<link rel='shortcut icon' href='//alpinelinux.org/alpine-logo.ico'/>
<link rel='alternate' title='Atom feed' href='https://git.alpinelinux.org/cgit/aports/atom/main/python3/bpo-30353.patch?h=master' type='application/atom+xml'/>
<link rel='vcs-git' href='git://git.alpinelinux.org/aports' title='aports Git repository'/>
<link rel='vcs-git' href='https://git.alpinelinux.org/cgit/aports' title='aports Git repository'/>
</head>
<body>
<div id='cgit'><table id='header'>
<tr>
<td class='logo' rowspan='2'><a href='/cgit/'><img src='//wiki.alpinelinux.org/w/resources/assets/alogo.png' alt='cgit logo'/></a></td>
<td class='main'><a href='/cgit/'>index</a> : <a title='aports' href='/cgit/aports/'>aports</a></td><td class='form'><form method='get'>
<select name='h' onchange='this.form.submit();'>
<option value='1.10-stable'>1.10-stable</option>
<option value='1.9'>1.9</option>
<option value='2.0-stable'>2.0-stable</option>
<option value='2.1-stable'>2.1-stable</option>
<option value='2.2-stable'>2.2-stable</option>
<option value='2.3-stable'>2.3-stable</option>
<option value='2.4-stable'>2.4-stable</option>
<option value='2.5-stable'>2.5-stable</option>
<option value='2.6-stable'>2.6-stable</option>
<option value='2.7-stable'>2.7-stable</option>
<option value='3.0-stable'>3.0-stable</option>
<option value='3.1-stable'>3.1-stable</option>
<option value='3.2-stable'>3.2-stable</option>
<option value='3.3-stable'>3.3-stable</option>
<option value='3.4-stable'>3.4-stable</option>
<option value='3.5-stable'>3.5-stable</option>
<option value='3.6-stable'>3.6-stable</option>
<option value='3.7-stable'>3.7-stable</option>
<option value='master' selected='selected'>master</option>
</select> <input type='submit' value='switch'/></form></td></tr>
<tr><td class='sub'>Main aports tree
</td><td class='sub right'>gitolite</td></tr></table>
<table class='tabs'><tr><td>
<a href='/cgit/aports/about/'>about</a><a href='/cgit/aports/'>summary</a><a href='/cgit/aports/refs/'>refs</a><a href='/cgit/aports/log/main/python3/bpo-30353.patch'>log</a><a class='active' href='/cgit/aports/tree/main/python3/bpo-30353.patch'>tree</a><a href='/cgit/aports/commit/main/python3/bpo-30353.patch'>commit</a><a href='/cgit/aports/diff/main/python3/bpo-30353.patch'>diff</a><a href='/cgit/aports/stats/main/python3/bpo-30353.patch'>stats</a></td><td class='form'><form class='right' method='get' action='/cgit/aports/log/main/python3/bpo-30353.patch'>
<select name='qt'>
<option value='grep'>log msg</option>
<option value='author'>author</option>
<option value='committer'>committer</option>
<option value='range'>range</option>
</select>
<input class='txt' type='text' size='10' name='q' value=''/>
<input type='submit' value='search'/>
</form>
</td></tr></table>
<div class='path'>path: <a href='/cgit/aports/tree/'>root</a>/<a href='/cgit/aports/tree/main'>main</a>/<a href='/cgit/aports/tree/main/python3'>python3</a>/<a href='/cgit/aports/tree/main/python3/bpo-30353.patch'>bpo-30353.patch</a></div><div class='content'>blob: 28fc93eef0b737e74e43dec0f148243f1c04aa7d (<a href='/cgit/aports/plain/main/python3/bpo-30353.patch'>plain</a>)
<table summary='blob content' class='blob'>
<tr><td class='linenumbers'><pre><a id='n1' href='#n1'>1</a>
<a id='n2' href='#n2'>2</a>
<a id='n3' href='#n3'>3</a>
<a id='n4' href='#n4'>4</a>
<a id='n5' href='#n5'>5</a>
<a id='n6' href='#n6'>6</a>
<a id='n7' href='#n7'>7</a>
<a id='n8' href='#n8'>8</a>
<a id='n9' href='#n9'>9</a>
<a id='n10' href='#n10'>10</a>
<a id='n11' href='#n11'>11</a>
<a id='n12' href='#n12'>12</a>
<a id='n13' href='#n13'>13</a>
<a id='n14' href='#n14'>14</a>
<a id='n15' href='#n15'>15</a>
<a id='n16' href='#n16'>16</a>
<a id='n17' href='#n17'>17</a>
<a id='n18' href='#n18'>18</a>
<a id='n19' href='#n19'>19</a>
<a id='n20' href='#n20'>20</a>
<a id='n21' href='#n21'>21</a>
<a id='n22' href='#n22'>22</a>
<a id='n23' href='#n23'>23</a>
<a id='n24' href='#n24'>24</a>
<a id='n25' href='#n25'>25</a>
<a id='n26' href='#n26'>26</a>
<a id='n27' href='#n27'>27</a>
<a id='n28' href='#n28'>28</a>
<a id='n29' href='#n29'>29</a>
<a id='n30' href='#n30'>30</a>
<a id='n31' href='#n31'>31</a>
<a id='n32' href='#n32'>32</a>
<a id='n33' href='#n33'>33</a>
<a id='n34' href='#n34'>34</a>
<a id='n35' href='#n35'>35</a>
<a id='n36' href='#n36'>36</a>
<a id='n37' href='#n37'>37</a>
<a id='n38' href='#n38'>38</a>
<a id='n39' href='#n39'>39</a>
<a id='n40' href='#n40'>40</a>
<a id='n41' href='#n41'>41</a>
<a id='n42' href='#n42'>42</a>
<a id='n43' href='#n43'>43</a>
<a id='n44' href='#n44'>44</a>
<a id='n45' href='#n45'>45</a>
<a id='n46' href='#n46'>46</a>
<a id='n47' href='#n47'>47</a>
<a id='n48' href='#n48'>48</a>
<a id='n49' href='#n49'>49</a>
<a id='n50' href='#n50'>50</a>
<a id='n51' href='#n51'>51</a>
<a id='n52' href='#n52'>52</a>
<a id='n53' href='#n53'>53</a>
<a id='n54' href='#n54'>54</a>
<a id='n55' href='#n55'>55</a>
<a id='n56' href='#n56'>56</a>
<a id='n57' href='#n57'>57</a>
<a id='n58' href='#n58'>58</a>
<a id='n59' href='#n59'>59</a>
<a id='n60' href='#n60'>60</a>
<a id='n61' href='#n61'>61</a>
<a id='n62' href='#n62'>62</a>
<a id='n63' href='#n63'>63</a>
<a id='n64' href='#n64'>64</a>
<a id='n65' href='#n65'>65</a>
<a id='n66' href='#n66'>66</a>
<a id='n67' href='#n67'>67</a>
<a id='n68' href='#n68'>68</a>
<a id='n69' href='#n69'>69</a>
<a id='n70' href='#n70'>70</a>
<a id='n71' href='#n71'>71</a>
<a id='n72' href='#n72'>72</a>
<a id='n73' href='#n73'>73</a>
<a id='n74' href='#n74'>74</a>
<a id='n75' href='#n75'>75</a>
<a id='n76' href='#n76'>76</a>
<a id='n77' href='#n77'>77</a>
<a id='n78' href='#n78'>78</a>
<a id='n79' href='#n79'>79</a>
<a id='n80' href='#n80'>80</a>
<a id='n81' href='#n81'>81</a>
<a id='n82' href='#n82'>82</a>
<a id='n83' href='#n83'>83</a>
<a id='n84' href='#n84'>84</a>
<a id='n85' href='#n85'>85</a>
<a id='n86' href='#n86'>86</a>
<a id='n87' href='#n87'>87</a>
<a id='n88' href='#n88'>88</a>
<a id='n89' href='#n89'>89</a>
<a id='n90' href='#n90'>90</a>
<a id='n91' href='#n91'>91</a>
<a id='n92' href='#n92'>92</a>
<a id='n93' href='#n93'>93</a>
<a id='n94' href='#n94'>94</a>
<a id='n95' href='#n95'>95</a>
<a id='n96' href='#n96'>96</a>
<a id='n97' href='#n97'>97</a>
<a id='n98' href='#n98'>98</a>
<a id='n99' href='#n99'>99</a>
<a id='n100' href='#n100'>100</a>
<a id='n101' href='#n101'>101</a>
<a id='n102' href='#n102'>102</a>
<a id='n103' href='#n103'>103</a>
<a id='n104' href='#n104'>104</a>
<a id='n105' href='#n105'>105</a>
<a id='n106' href='#n106'>106</a>
<a id='n107' href='#n107'>107</a>
<a id='n108' href='#n108'>108</a>
<a id='n109' href='#n109'>109</a>
<a id='n110' href='#n110'>110</a>
<a id='n111' href='#n111'>111</a>
<a id='n112' href='#n112'>112</a>
<a id='n113' href='#n113'>113</a>
<a id='n114' href='#n114'>114</a>
<a id='n115' href='#n115'>115</a>
<a id='n116' href='#n116'>116</a>
<a id='n117' href='#n117'>117</a>
<a id='n118' href='#n118'>118</a>
<a id='n119' href='#n119'>119</a>
<a id='n120' href='#n120'>120</a>
<a id='n121' href='#n121'>121</a>
<a id='n122' href='#n122'>122</a>
<a id='n123' href='#n123'>123</a>
<a id='n124' href='#n124'>124</a>
<a id='n125' href='#n125'>125</a>
<a id='n126' href='#n126'>126</a>
<a id='n127' href='#n127'>127</a>
<a id='n128' href='#n128'>128</a>
<a id='n129' href='#n129'>129</a>
<a id='n130' href='#n130'>130</a>
<a id='n131' href='#n131'>131</a>
<a id='n132' href='#n132'>132</a>
<a id='n133' href='#n133'>133</a>
</pre></td>
<td class='lines'><pre><code>https://github.com/python/cpython/commit/9ba3aa4d02a110d1a1ea464a8aff3be7dd9c63c3.patch

From 9ba3aa4d02a110d1a1ea464a8aff3be7dd9c63c3 Mon Sep 17 00:00:00 2001
From: Erik Bray &lt;erik.m.bray&#64;gmail.com&gt;
Date: Wed, 7 Jun 2017 18:42:24 +0100
Subject: [PATCH] bpo-30353: Fix pass by value for structs on 64-bit
 Cygwin/MinGW (GH-1559)

<span style="color:#0057ae">---</span>
 Lib/ctypes/test/test_as_parameter.py |  4 ++++
 Lib/ctypes/test/test_structures.py   | 22 ++++++++++++++++++++++
 Modules/_ctypes/_ctypes_test.c       | 18 ++++++++++++++++++
 Modules/_ctypes/callproc.c           | 23 +++++++++++++++++++++--
 4 files changed, 65 insertions(+), 2 deletions(-)

diff --git a/Lib/ctypes/test/test_as_parameter.py b/Lib/ctypes/test/test_as_parameter.py
index 2a3484bec01..a2640575a07 100644
<span style="color:#0057ae">--- a/Lib/ctypes/test/test_as_parameter.py</span>
<span style="color:#000000; font-weight:bold">+++ b/Lib/ctypes/test/test_as_parameter.py</span>
&#64;&#64; -169,6 +169,10 &#64;&#64; class S2H(Structure):
         s2h = dll.ret_2h_func(self.wrap(inp))
         self.assertEqual((s2h.x, s2h.y), (99*2, 88*3))
 
<span style="color:#000000; font-weight:bold">+        # Test also that the original struct was unmodified (i.e. was passed by</span>
<span style="color:#000000; font-weight:bold">+        # value)</span>
<span style="color:#000000; font-weight:bold">+        self.assertEqual((inp.x, inp.y), (99, 88))</span>
<span style="color:#000000; font-weight:bold">+</span>
     def test_struct_return_8H(self):
         class S8I(Structure):
             _fields_ = [(&quot;a&quot;, c_int),
diff --git a/Lib/ctypes/test/test_structures.py b/Lib/ctypes/test/test_structures.py
index 2e778fb1b43..d90c71144c9 100644
<span style="color:#0057ae">--- a/Lib/ctypes/test/test_structures.py</span>
<span style="color:#000000; font-weight:bold">+++ b/Lib/ctypes/test/test_structures.py</span>
&#64;&#64; -417,6 +417,28 &#64;&#64; class X(Structure):
         self.assertEqual(s.second, 0xcafebabe)
         self.assertEqual(s.third, 0x0bad1dea)
 
<span style="color:#000000; font-weight:bold">+    def test_pass_by_value_in_register(self):</span>
<span style="color:#000000; font-weight:bold">+        class X(Structure):</span>
<span style="color:#000000; font-weight:bold">+            _fields_ = [</span>
<span style="color:#000000; font-weight:bold">+                (&apos;first&apos;, c_uint),</span>
<span style="color:#000000; font-weight:bold">+                (&apos;second&apos;, c_uint)</span>
<span style="color:#000000; font-weight:bold">+            ]</span>
<span style="color:#000000; font-weight:bold">+</span>
<span style="color:#000000; font-weight:bold">+        s = X()</span>
<span style="color:#000000; font-weight:bold">+        s.first = 0xdeadbeef</span>
<span style="color:#000000; font-weight:bold">+        s.second = 0xcafebabe</span>
<span style="color:#000000; font-weight:bold">+        dll = CDLL(_ctypes_test.__file__)</span>
<span style="color:#000000; font-weight:bold">+        func = dll._testfunc_reg_struct_update_value</span>
<span style="color:#000000; font-weight:bold">+        func.argtypes = (X,)</span>
<span style="color:#000000; font-weight:bold">+        func.restype = None</span>
<span style="color:#000000; font-weight:bold">+        func(s)</span>
<span style="color:#000000; font-weight:bold">+        self.assertEqual(s.first, 0xdeadbeef)</span>
<span style="color:#000000; font-weight:bold">+        self.assertEqual(s.second, 0xcafebabe)</span>
<span style="color:#000000; font-weight:bold">+        got = X.in_dll(dll, &quot;last_tfrsuv_arg&quot;)</span>
<span style="color:#000000; font-weight:bold">+        self.assertEqual(s.first, got.first)</span>
<span style="color:#000000; font-weight:bold">+        self.assertEqual(s.second, got.second)</span>
<span style="color:#000000; font-weight:bold">+</span>
<span style="color:#000000; font-weight:bold">+</span>
 class PointerMemberTestCase(unittest.TestCase):
 
     def test(self):
diff --git a/Modules/_ctypes/_ctypes_test.c b/Modules/_ctypes/_ctypes_test.c
index fe0015c8013..2255e573393 100644
<span style="color:#0057ae">--- a/Modules/_ctypes/_ctypes_test.c</span>
<span style="color:#000000; font-weight:bold">+++ b/Modules/_ctypes/_ctypes_test.c</span>
&#64;&#64; -57,6 +57,24 &#64;&#64; _testfunc_large_struct_update_value(Test in)
     ((volatile Test *)&amp;in)-&gt;third = 0x0badf00d;
 }
 
<span style="color:#000000; font-weight:bold">+typedef struct {</span>
<span style="color:#000000; font-weight:bold">+    unsigned int first;</span>
<span style="color:#000000; font-weight:bold">+    unsigned int second;</span>
<span style="color:#000000; font-weight:bold">+} TestReg;</span>
<span style="color:#000000; font-weight:bold">+</span>
<span style="color:#000000; font-weight:bold">+</span>
<span style="color:#000000; font-weight:bold">+EXPORT(TestReg) last_tfrsuv_arg;</span>
<span style="color:#000000; font-weight:bold">+</span>
<span style="color:#000000; font-weight:bold">+</span>
<span style="color:#000000; font-weight:bold">+EXPORT(void)</span>
<span style="color:#000000; font-weight:bold">+_testfunc_reg_struct_update_value(TestReg in)</span>
<span style="color:#000000; font-weight:bold">+{</span>
<span style="color:#000000; font-weight:bold">+    last_tfrsuv_arg = in;</span>
<span style="color:#000000; font-weight:bold">+    ((volatile TestReg *)&amp;in)-&gt;first = 0x0badf00d;</span>
<span style="color:#000000; font-weight:bold">+    ((volatile TestReg *)&amp;in)-&gt;second = 0x0badf00d;</span>
<span style="color:#000000; font-weight:bold">+}</span>
<span style="color:#000000; font-weight:bold">+</span>
<span style="color:#000000; font-weight:bold">+</span>
 EXPORT(void)testfunc_array(int values[4])
 {
     printf(&quot;testfunc_array %d %d %d %d\n&quot;,
diff --git a/Modules/_ctypes/callproc.c b/Modules/_ctypes/callproc.c
index 0b6faf96c68..5439b939dc4 100644
<span style="color:#0057ae">--- a/Modules/_ctypes/callproc.c</span>
<span style="color:#000000; font-weight:bold">+++ b/Modules/_ctypes/callproc.c</span>
&#64;&#64; -1039,6 +1039,13 &#64;&#64; GetComError(HRESULT errcode, GUID *riid, IUnknown *pIunk)
 }
 #endif
 
<span style="color:#000000; font-weight:bold">+#if (defined(__x86_64__) &amp;&amp; (defined(__MINGW64__) || defined(__CYGWIN__))) || \</span>
<span style="color:#000000; font-weight:bold">+    defined(__aarch64__)</span>
<span style="color:#000000; font-weight:bold">+#define CTYPES_PASS_BY_REF_HACK</span>
<span style="color:#000000; font-weight:bold">+#define POW2(x) (((x &amp; ~(x - 1)) == x) ? x : 0)</span>
<span style="color:#000000; font-weight:bold">+#define IS_PASS_BY_REF(x) (x &gt; 8 || !POW2(x))</span>
<span style="color:#000000; font-weight:bold">+#endif</span>
<span style="color:#000000; font-weight:bold">+</span>
 /*
  * Requirements, must be ensured by the caller:
  * - argtuple is tuple of arguments
&#64;&#64; -1136,8 +1143,20 &#64;&#64; PyObject *_ctypes_callproc(PPROC pProc,
     }
     for (i = 0; i &lt; argcount; ++i) {
         atypes[i] = args[i].ffi_type;
<span style="color:#0057ae">-        if (atypes[i]-&gt;type == FFI_TYPE_STRUCT</span>
<span style="color:#0057ae">-            )</span>
<span style="color:#000000; font-weight:bold">+#ifdef CTYPES_PASS_BY_REF_HACK</span>
<span style="color:#000000; font-weight:bold">+        size_t size = atypes[i]-&gt;size;</span>
<span style="color:#000000; font-weight:bold">+        if (IS_PASS_BY_REF(size)) {</span>
<span style="color:#000000; font-weight:bold">+            void *tmp = alloca(size);</span>
<span style="color:#000000; font-weight:bold">+            if (atypes[i]-&gt;type == FFI_TYPE_STRUCT)</span>
<span style="color:#000000; font-weight:bold">+                memcpy(tmp, args[i].value.p, size);</span>
<span style="color:#000000; font-weight:bold">+            else</span>
<span style="color:#000000; font-weight:bold">+                memcpy(tmp, (void*)&amp;args[i].value, size);</span>
<span style="color:#000000; font-weight:bold">+</span>
<span style="color:#000000; font-weight:bold">+            avalues[i] = tmp;</span>
<span style="color:#000000; font-weight:bold">+        }</span>
<span style="color:#000000; font-weight:bold">+        else</span>
<span style="color:#000000; font-weight:bold">+#endif</span>
<span style="color:#000000; font-weight:bold">+        if (atypes[i]-&gt;type == FFI_TYPE_STRUCT)</span>
             avalues[i] = (void *)args[i].value.p;
         else
             avalues[i] = (void *)&amp;args[i].value;
</code></pre></td></tr></table>
</div> <!-- class=content -->
<script type="text/javascript">
var gaJsHost = (("https:" == document.location.protocol) ? "https://ssl." : "http://www.");
document.write(unescape("%3Cscript src='" + gaJsHost + "google-analytics.com/ga.js' type='text/javascript'%3E%3C/script%3E"));
</script>
<script type="text/javascript">
try {
var pageTracker = _gat._getTracker("UA-2804446-12");
pageTracker._trackPageview();
} catch(err) {}</script>
</div> <!-- id=cgit -->
</body>
</html>
